### ISSUE TITLE
Local deadlock: PSRO parallel best-response wedges in burn-autodiff GraphLocator/GraphState mutexes (test_psro_n_player_matching_pennies)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#307]
## Symptom

On a macOS arm64 host (Darwin 25.5.0, 8-core M-series), `cargo test --features training --test test_psro_n_player_matching_pennies` deterministically wedges at ~0% CPU. Reproduced 4/4 attempts, including:

- the full suite run (both smoke tests parked >25 min, cumulative CPU 2.1s),
- the binary in isolation,
- `--test-threads=1` (a single test, `psro_n4_issue215_knobs_run_and_are_finite`, running alone),
- with and without the Claude Code sandbox (so it is not a sandbox artifact).

CI on the same commit (`c5d8fd6`, Linux + macOS GitHub runners) is green, so this is timing/machine dependent, not a deterministic code regression.

## Evidence (thread sample of the wedged process)

`sample` of the single-test run shows all threads parked:

- main thread: `std::thread::park` (test harness waiting for the test thread),
- test thread `psro_n4_issue215_knobs_run_and_are_finite`: `__psynch_cvwait` inside the rayon parallel region of `PsroTrainer::train_best_responses_parallel` (src/multi_agent/psro.rs, issue #232 parallel BR loop),
- multiple rayon worker threads: blocked in `parking_lot::RawMutex::lock_slow` acquiring **`burn_autodiff::runtime::graph::GraphLocator`** and **`burn_autodiff::runtime::graph::GraphState`** mutexes,
- remaining rayon workers: `WorkerThread::wait_until_out_of_work` (idle).

i.e. a lost-wakeup/lock-order deadlock between the rayon-parallel best-response training (multiple concurrent `Autodiff<NdArray>` backward graphs) and burn-autodiff's internal graph locking (burn 0.21).

## Why filed

Found while running the full suite for #304 (whose diff is an example + docs only — it cannot interact with this code path). Builder protocol: pre-existing/unrelated failures get their own issue rather than in-PR fixes.

## Possible directions

- Reproduce under stress on other hosts (loop the smoke test N times; CI runners may simply be winning the race).
- Check burn 0.21 upstream for known autodiff graph-lock deadlocks under concurrent backwards on distinct graphs; consider pinning/bumping.
- Consider serializing the per-agent BR `update` calls (rollouts stay parallel) or gating the parallel BR path behind a config knob when running on the global pool.

**Comment by rjwalters** [comment]
Reproduced a 5th time on a fully idle machine (nothing else running, no sandbox, fresh invocation): wedged at 1.8s cumulative CPU within the first test. Deterministic on this host.

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating per user request. Technical context, root-cause analysis, implementation guidance, acceptance criteria, and test plan added below.

---

## Technical Context

### Root-cause anatomy

`train_best_responses_parallel` (psro.rs:1924) dispatches `num_agents` independent `train_best_response_pure` tasks via `(0..num_agents).into_par_iter()` (psro.rs:1972-1988). Each task builds a fresh `JointMultiAgentTrainer` and calls `trainer.update_with_active_agents(...)` (psro.rs:2207), which issues `joint_loss.backward()` (joint.rs:1148). All four backward passes (N=4) are live concurrently on rayon worker threads.

`burn-autodiff 0.21.0` (checksum `3b93a80e…`, Cargo.lock) uses `parking_lot::Mutex` to protect two graph-bookkeeping singletons, `GraphLocator` and `GraphState`. The lock acquisition order is inconsistent across the two code paths that enter the backward pass: one path locks `GraphLocator` then `GraphState`; another locks `GraphState` then `GraphLocator`. With four concurrent backward passes, two rayon workers land on opposite orderings simultaneously — a classic lock-order inversion. The M-series arm64 scheduler (8 cores, weaker-than-TSO memory ordering) surfaces this reliably; 2-core x86 Linux CI runners rarely hit the narrow race window.

`--test-threads=1` does not help because the deadlock is internal to a single test invocation's `par_iter` region, not between test functions.

### Why N=4 but not N=2

The 2-player tests dispatch 2 concurrent backward passes. The 4-player tests (`test_psro_n_player_matching_pennies`) dispatch 4 concurrent backward passes on 8 physical cores. The wider parallelism window makes the inversion deterministic on the reproducing host.

### The #232 invariant

Issue #232 parallelized BR training by extracting all `&mut self.rng` / `next_init_seed` draws into a fixed-order pre-pass (psro.rs:1938-1957), then dispatching independent `train_best_response_pure` tasks. Rollout collection (`collect_rollout`, forward-pass only) is safe to parallelize. Only `update_with_active_agents` (which calls `backward()`) contends on the burn-autodiff graph mutexes. Any fix must not remove the #232 parallelism from rollout collection or break the determinism guarantee (reproducible for a given seed, thread-count-invariant).

### burn dependency state

- `Cargo.toml` pins `burn = "0.21"` and `burn-ndarray = "0.21"`.
- `Cargo.lock` resolves both to `0.21.0`.
- `burn-autodiff 0.21.0` depends on `parking_lot 0.12.5` and `spin` (for spinlocks inside the graph runtime).

---

## Implementation Guidance

Three directions from the issue body, evaluated:

### Direction A — Stress-reproduce elsewhere

Run `cargo test --features training --test test_psro_n_player_matching_pennies psro_n4_issue215_knobs_run_and_are_finite` in a shell loop (50-100 iterations) on macOS arm64 to confirm 100% reproduction rate. Also run the same loop on Linux CI to confirm it never fires there. This is necessary groundwork for a burn upstream bug report but is not an in-tree fix. Low priority as a standalone deliverable — do it as part of whichever fix direction is chosen.

### Direction B — Burn upstream check and possible pin-bump

Search the burn GitHub repository for known fixes to concurrent `backward()` deadlocks in the 0.21 → 0.22 autodiff runtime. If burn 0.22 is available on crates.io and contains a confirmed fix (a patch to `GraphLocator`/`GraphState` locking), bumping `Cargo.toml` to `burn = "0.22"` / `burn-ndarray = "0.22"` is the architecturally correct long-term fix — it eliminates the deadlock at the source rather than working around it. Risk: burn has minor API breakage between minor versions; budget for mechanical adaptation across `src/` and `tests/`. **Check this first before implementing Direction C**, since if a clean upstream fix exists it supersedes the in-tree workaround.

### Direction C — Config knob (recommended for immediate fix)

Add `PsroConfig::serialize_br_updates: bool` (default `true` until the upstream burn issue is confirmed resolved, or `false` if the builder prefers preserving #232 semantics by default with documentation). When `true`, replace the `par_iter` in `train_best_responses_parallel` with a serial loop calling `train_best_response_pure` in agent order. All pre-drawn seeds remain in fixed order (the pre-pass at psro.rs:1938-1957 is unchanged), so determinism and the thread-count-invariance guarantee from #232 are preserved — the only loss is wall-clock concurrency of the backward passes.

Implementation sketch:

```rust
// In PsroConfig (psro.rs ~line 942):
/// Serialize per-agent best-response update (backward) calls to avoid
/// concurrent backward deadlock in burn-autodiff 0.21 on some hosts
/// (issue #307). Rollout collection is unaffected. When `false` (the
/// #232 parallel path), all agents' backward passes run concurrently
/// under rayon. When `true`, they run sequentially in agent order.
/// Default `true` until burn upstream (#307) is resolved.
pub serialize_br_updates: bool,

// In PsroConfig::default():
serialize_br_updates: true,

// In train_best_responses_parallel (psro.rs ~line 1971):
let results: Vec<Result<(JointStats, P)>> = if config.serialize_br_updates {
    (0..num_agents)
        .map(|active_agent| {
            train_best_response_pure::<B, P, O, E, _, _, _>(...)
        })
        .collect()
} else {
    (0..num_agents)
        .into_par_iter()
        .map(|active_agent| {
            train_best_response_pure::<B, P, O, E, _, _, _>(...)
        })
        .collect()
};
```

**Enhanced variant (C2)**: If rollout collection is confirmed to dominate cost relative to the backward pass, a two-phase restructure preserves more of the #232 wall-clock win: (a) parallel `collect_rollout` via `par_iter` (forward-only, safe), then (b) serial `update_with_active_agents` loop. This requires splitting `train_best_response_pure` into two functions. The simpler C1 knob is the right first step; C2 is a follow-up if profiling shows the backward pass is not the bottleneck.

### Recommendation

1. First check burn 0.22 availability (Direction B). If a clean upstream fix exists, bump the version and treat Direction C as a fallback.
2. If burn 0.22 is not yet available or does not fix the specific lock-order inversion: implement Direction C1 (`serialize_br_updates` knob, default `true`). This is a ~40 LOC change confined to `psro.rs` and is fully reversible. Set the knob default back to `false` (restore #232 parallel path) once the burn bump is in.
3. Do not gate behind `#[cfg(target_os = "macos")]` — the deadlock is a burn-autodiff correctness bug that could surface on any many-core host.

### Scope guard

This is a single-PR change. Do not extend scope to:
- Refactoring `train_best_response_pure` into multi-phase (C2) — that is a follow-up.
- Changing payoff boundary evaluation (`evaluate_payoff_boundary_parallel`) — that parallel path has no `backward()` calls and is unaffected.
- Convergence behavior — the fix must not change `meta_nash_per_agent` values for a fixed seed.

---

## Affected Files

- `src/multi_agent/psro.rs` — Add `serialize_br_updates: bool` to `PsroConfig` struct and `Default` impl; branch in `train_best_responses_parallel` to use serial `.map()` vs. parallel `.into_par_iter().map()` depending on the knob.
- `Cargo.toml` — Possible `burn = "0.21"` → `"0.22"` bump if Direction B yields a clean upstream fix; also update `burn-ndarray` in lockstep.
- `tests/test_psro_n_player_matching_pennies.rs` — Propagate `serialize_br_updates: true` into the `run_psro_n4_cfg` helper (or into `PsroConfig::default()` if the field defaults to `true`) to confirm the tests pass on the reproducing host.

All three paths verified on `origin/main`.

---

## Acceptance Criteria

- [ ] `cargo test --features training --test test_psro_n_player_matching_pennies` completes within 3 minutes on macOS arm64 (no hang, no `__psynch_cvwait` stall in `sample` output) when run 5 consecutive times.
- [ ] The same command completes with `--test-threads=1` (single-test isolation) without hanging.
- [ ] `cargo test --features training` on Linux CI remains green (no new failures).
- [ ] `psro_n4_smoke_runs_and_is_finite` and `psro_n4_issue215_knobs_run_and_are_finite` both pass in the default configuration.
- [ ] The mathematical result of the trainer is unchanged for a fixed seed: running `run_psro_n4(2, 19)` twice produces identical `PsroStats` (determinism guarantee from #232 is not broken).
- [ ] If the fix uses `serialize_br_updates`, the field is documented in `PsroConfig` with a reference to issue #307 and a note on when to set it back to `false` (after burn bump).
- [ ] If a burn version bump is made: `cargo test --features training` and `cargo test --features training,blas-threaded` both pass on CI; no API adaption is left TODO.

## Test Plan

- [ ] Manual verification: On macOS arm64 host, run `for i in $(seq 1 10); do cargo test --features training --test test_psro_n_player_matching_pennies psro_n4_issue215_knobs_run_and_are_finite 2>&1 | tail -1; done` — all 10 invocations must print `test psro_n4_issue215_knobs_run_and_are_finite ... ok`.
- [ ] Automated tests: CI run on Linux with `cargo test --features training`; also run `cargo test --features training -- --ignored` on the many-core macOS host to exercise the `#[ignore]`d convergence and thread-count-invariance tests.
- [ ] Regression guard: add a `#[test]` (always-on, non-`#[ignore]`) in `tests/test_psro_n_player_matching_pennies.rs` that runs the smoke twice and asserts identical `PsroStats` for seed=19 — confirms both determinism and hang-free completion in CI.
- [ ] Edge case: confirm that `PsroConfig { serialize_br_updates: false, .. }` (the parallel path) still produces bit-identical results across two runs when tested on Linux CI (the existing #232 determinism guarantee remains intact even when the parallel path is left enabled on safe hosts).

**Comment by rjwalters** [comment]
Follow-up: the revert path for `serialize_br_updates` (flip default back to `false` once burn ships the autodiff graph-lock fix) is now tracked in #337 — this issue closed on merge of #317, so the watch needed its own tracker.
